File: 3/solution.rs
fn round_sqrt(n: u64) -> u64 {
    (n as f64).sqrt() as u64 + 1
}

fn is_prime(n: u64) -> bool {
    if (n%2 == 0) & (n > 2) {
        return false;
    } else {
        (3..round_sqrt(n)).all(|x| n % x != 0)
    }
}

fn largest_factor(n: u64) -> u64 {
    (1..(round_sqrt(n))).filter(|&x| (n%x == 0) & is_prime(x)).max().unwrap_or(0)
}

fn main() {
    assert!(largest_factor(13195) == 29);
    let result = largest_factor(600851475143);
    println!("{}", result.to_string())
}

File: 3/solution.hs
sieve :: Integral t => [t] -> [t]
sieve [] = []
sieve (p:xs) = p : sieve [x | x <- xs, x `mod` p > 0]

primes :: [Int]
primes = sieve [2..]

primeFactors :: Int -> [Int]
primeFactors n = [x | x <- takeWhile (< n) primes, n `mod` x == 0]

-- Coudn't get sieve version fast enough???

floorRoot = floor . sqrt . fromIntegral

isPrime n
  | n `mod` 2 == 0 && n > 2 = False
  | otherwise = all (\x -> n `mod` x /= 0) [x | x <- [3,5..(floor . sqrt . fromIntegral) n]]

pFactors n = filter (\x -> n `mod` x == 0 && isPrime x) [1..(floorRoot n)]

main::IO ()
main = print $ maximum (pFactors 600851475143)
